### PR TITLE
support adamw optimizer and part optimizer and label soomthing

### PR DESCRIPTION
**Pull request by chengaofei**
support adamw optimizer

### CODE REVIEW

File: tzrec/optim/optimizer_builder.py [hunk 1]
```diff
@@ -115,3 +140,88 @@ def create_scheduler(
...
     # pyre-ignore [16]
     return BaseLR.create_class(lr_cls_name)(**lr_kwargs)
+
+
+def create_part_optim_schedulers(
+    part_optimizers: List[KeyedOptimizerWrapper],
+    optimizer_config: optimizer_pb2.DenseOptimizer,
+    part_optim_indices: List[int],
+) -> List[BaseLR]:
+    """Create optimizer for dense module.
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py]: Create scheduler for part of module parameters.

File: tzrec/optim/optimizer_builder.py [hunk 2]
```diff
@@ -88,10 +93,30 @@ def create_dense_optimizer(
...
+    part_optimizer_kwargs = []
+    part_regex_patterns = []
+    for part_optimizer in optimizer_config.part_optimizers:
+        optim, kwargs = create_dense_optimizer(part_optimizer)
+        part_optimizers.append(optim)
+        part_optimizer_kwargs.append(kwargs)
+        part_regex_patterns.append(part_optimizer.regex_pattern)
+    return part_optimizers, part_optimizer_kwargs, part_regex_patterns
+
+
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py:119]: Create scheduler for dense module.

File: tzrec/optim/optimizer_builder.py [hunk 3]
```diff
@@ -115,3 +140,88 @@ def create_scheduler(
...
+def group_param_by_regex_pattern(
+    params: Dict[str, nn.Parameter], regex_patterns: list[str]
+) -> Tuple[Dict[str, nn.Parameter], List[Dict[str, nn.Parameter]]]:
+    """Group params by regex."""
+    remaining_params = dict()
+    part_optim_params = [dict() for _ in range(len(regex_patterns))]
+    for name, param in params.items():
+        logger.info("parameter name: {}".format(name))
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py]: remove the info

File: tzrec/optim/optimizer_builder.py [hunk 4]
```diff
@@ -115,3 +140,88 @@ def create_scheduler(
...
+        for i, regex_pattern in enumerate(regex_patterns):
+            if re.fullmatch(re.compile(regex_pattern), name):
+                part_optim_params[i][name] = param
+                break
+        else:
+            remaining_params[name] = param
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py:197]:
```
if len(part_optim_params) > 0:
    logger.info part_optim_params
    logger.info remaining_params
```

File: tzrec/main.py [hunk 5]
```diff
@@ -636,18 +636,31 @@ def train_and_evaluate(
...
+    part_optim_cls, part_optim_kwargs, part_optim_regex_patterns = (
+        optimizer_builder.create_part_optimizer(train_config.dense_optimizer)
+    )
+    remaining_params, part_optim_params = (
+        optimizer_builder.group_param_by_regex_pattern(
+            dict(in_backward_optimizer_filter(model.named_parameters())),
+            part_optim_regex_patterns,
+        )
+    )
+    logger.info("remaining_params:" + str(list(remaining_params.keys())))
```
Comment (tiankongdeguiji) [inline tzrec/main.py]: remove the info

File: tzrec/optim/optimizer_builder.py [hunk 6]
```diff
@@ -88,10 +93,30 @@ def create_dense_optimizer(
...
+        optimizer_kwargs["betas"] = (beta1, beta2)
+        return torch.optim.AdamW, optimizer_kwargs
     else:
         raise ValueError(f"Unknown optimizer: {optimizer_type}")
 
 
+def create_part_optimizer(
+    optimizer_config: optimizer_pb2.DenseOptimizer,
+) -> Tuple[List[Type[Optimizer]], List[Dict[str, Any]], List[str]]:
+    """Create optimizer class for some param."""
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py]: Create optimizer for part of module parameters.

File: tzrec/optim/optimizer_builder.py [hunk 7]
```diff
@@ -115,3 +140,88 @@ def create_scheduler(
...
+    optimizer_config: optimizer_pb2.DenseOptimizer,
+    part_optim_indices: List[int],
+) -> List[BaseLR]:
+    """Create optimizer for dense module.
+
+    Args:
+        part_optimizers (list[Optimizer]): an list instance of Optimizer.
+        optimizer_config (optimizer_pb2.SparseOptimizer|optimizer_pb2.DenseOptimizer):
+            an instance of Optimizer config.
+        part_optim_indices (list[int]): optimizers index in
```
Comment (tiankongdeguiji) [inline tzrec/optim/optimizer_builder.py]: why we need part_optim_indices? directly use optimizer_config.part_optimizers

File: tzrec/protos/optimizer.proto [hunk 8]
```diff
@@ -24,12 +24,29 @@ message DenseOptimizer {
...
     }
     oneof learning_rate {
         ConstantLR constant_learning_rate = 101;
         ExponentialDecayLR exponential_decay_learning_rate = 102;
         ManualStepLR manual_step_learning_rate = 103;
     }
+    repeated PartOptimizer part_optimizers = 201;
+}
+
+message PartOptimizer {
```
Comment (tiankongdeguiji) [inline tzrec/protos/optimizer.proto:37]: add a document in torcheasyrec.readthedocs.io

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/optim/optimizer_builder.py (hunk 7 above)
Comment (tiankongdeguiji) [reply tzrec/optim/optimizer_builder.py]: valid optimizers index in optimizer_config.part_optimizers.

File: tzrec/main.py (hunk 5 above)
Comment (chengaofei) [reply tzrec/main.py]: fix

File: tzrec/optim/optimizer_builder.py (hunk 6 above)
Comment (chengaofei) [reply tzrec/optim/optimizer_builder.py]: fix

File: tzrec/optim/optimizer_builder.py (hunk 2 above)
Comment (chengaofei) [reply tzrec/optim/optimizer_builder.py:119]: fix

File: tzrec/optim/optimizer_builder.py (hunk 1 above)
Comment (chengaofei) [reply tzrec/optim/optimizer_builder.py]: fix

File: tzrec/optim/optimizer_builder.py (hunk 3 above)
Comment (chengaofei) [reply tzrec/optim/optimizer_builder.py]: fix

File: tzrec/optim/optimizer_builder.py (hunk 4 above)
Comment (chengaofei) [reply tzrec/optim/optimizer_builder.py:197]: fix

File: docs/source/models/optimizer.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# 优化器
+
+## 简介
+
+TorchEasyRec除了像tzrec一样拥有sparse_optimizer和dense_optimizer,在dense_optimize中实现了part_optimizers 即：参数优化器。对于模型参数优先part_optimizer的正则项和参数名称匹配，如果能匹配上则该参数使用对应的part_optimizer，如果最后无法匹配上，则使用全局的dense_optimizer。
```
Comment (tiankongdeguiji) [inline docs/source/models/optimizer.md]: TorchEasyRec的优化器分为sparse_optimizer和dense_optimizer两个部分，sparse_optimizer负责embedding部分稀疏参数，dense_optimizer负责nn部分稠密参数。在dense_optimizer中，TorchEasyRec还支持part_optimizers，可以通过正则表示支持对部分参数设置单独的优化器。

File: docs/source/models/optimizer.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                decay_size: [1000, 10000]
+                learning_rates:[0.001, 0.0001]
+            }
+        }
+
+    }
+}
+```
+
+如上，在train_config.dense_optimizer中的part_optimizers是参数优化器
```
Comment (tiankongdeguiji) [inline docs/source/models/optimizer.md]: 在train_config.dense_optimizer中可以通过part_optimizers针对部分稠密参数配置单独的优化器

File: docs/source/models/optimizer.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+- optimizer: 和dense_optimizer可配置项一样，具体见proto的定义
```
Comment (tiankongdeguiji) [inline docs/source/models/optimizer.md]:
```
具体见optimize的[配置文档](reference.html)
```

File: docs/source/models/optimizer.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            regex_pattern: "(.*)light_mlp(.*)"
+            exponential_decay_learning_rate {
+                decay_size: [1000, 10000]
+                learning_rates:[0.001, 0.0001]
+            }
+        }
+
+    }
+}
+```
```
Comment (tiankongdeguiji) [inline docs/source/models/optimizer.md:43]: 增加sparse_optimizer和dense_optimizer的说明文档

File: tzrec/optim/optimizer_builder.py (hunk 4 above)
Comment (tiankongdeguiji) [reply tzrec/optim/optimizer_builder.py:197]: do not logging remaining params when len(part_optim_params) == 0
